pymc3/model.py

```diff
@@ -1054,6 +1054,113 @@ def _repr_latex_(self, name=None, dist=None):
 
     __latex__ = _repr_latex_
 
+    def predict(self, trace, dict_new_values, **kwargs):
+        """Predicts output data given a new set of input data.
+
+        Parameters
+        ----------
+        trace : Trace
+        dict_new_values : dict
+            New values for the data containers. The keys of the dictionary are
+            the  variables names in the model and the values are the objects
+            with which to update.
+            Before returning from the function, the variables return to their
+            original values.
+        **kwargs : keyword arguments
+            Extra keyword arguments are forwarded to
+            pymc3.sample_posterior_predictive.
+
+        Returns
+        -------
+        ppc : dict
+            Output of the function pymc3.sample_posterior_predictive() using
+            new values for the data containers.
+            Dictionary with the variables as keys. The values corresponding to
+            the posterior predictive samples from the weighted models.
+
+        Examples
+        --------
+
+        .. code:: ipython
+
+            >>> import pymc3 as pm
+            >>> with pm.Model() as model:
+            ...    x = pm.Data('x', [1., 2., 3.])
+            ...    y = pm.Data('y', [1., 2., 3.])
+            ...    beta = pm.Normal('beta', 0, 1)
+            ...    obs = pm.Normal('obs', x * beta, 1, observed=y)
+            ...    trace = pm.sample(1000, tune=1000)
+
+        .. code:: ipython
+
+            >>> # Predict on new data.
+            >>> y_test = model.predict(trace, {'x': [5,6,9]})
+            >>> y_test['obs'].mean(axis=0)
+            array([4.6088569 , 5.54128318, 8.32953844])
+        """
+        dict_original_values = {}
+        for variable_name, new_value in dict_new_values.items():
+            original_value = self[variable_name].get_value()
+            dict_original_values[variable_name] = original_value
+            self[variable_name].set_value(pandas_to_array(new_value))
+
+        ppc = pm.sample_posterior_predictive(trace=trace, model=self, **kwargs)
+
+        for variable_name, original_value in dict_original_values.items():
```

Comment:
cool -- so calling this method does not change the state of the `Data` or `model`. 

I think this should be wrapped in a `try/finally`, since we tend to try to handle a `StopExecution` during sampling?

```python
try:
    ppc = pm.sample_posterior_predictive(...)
finally:
    for variable_name, original_value in dict_original_values.items():
        ...
```